executor/aggfuncs/func_max_min.go

```diff
@@ -14,122 +14,135 @@
 package aggfuncs
 
 import (
-	"container/heap"
 	"unsafe"
 
+	"github.com/pingcap/errors"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/stringutil"
 )
 
-type maxMinHeap struct {
-	data    []interface{}
-	h       heap.Interface
-	varSet  map[interface{}]int64
-	isMax   bool
-	cmpFunc func(i, j interface{}) int
+// newDeque inits a new Deque
+func newDeque(isMax bool, cmpFunc func(i, j interface{}) int) *Deque {
+	return &Deque{[]Pair{}, isMax, cmpFunc}
 }
 
-func newMaxMinHeap(isMax bool, cmpFunc func(i, j interface{}) int) *maxMinHeap {
-	h := &maxMinHeap{
-		data:    make([]interface{}, 0),
-		varSet:  make(map[interface{}]int64),
-		isMax:   isMax,
-		cmpFunc: cmpFunc,
-	}
-	return h
+// Deque is an array based double end queue
+type Deque struct {
+	Items   []Pair
+	IsMax   bool
+	cmpFunc func(i, j interface{}) int
 }
 
-func (h *maxMinHeap) Len() int { return len(h.data) }
-func (h *maxMinHeap) Less(i, j int) bool {
-	if h.isMax {
-		return h.cmpFunc(h.data[i], h.data[j]) > 0
-	}
-	return h.cmpFunc(h.data[i], h.data[j]) < 0
+// PushFront pushes idx and item(wrapped in Pair) to the front of Deque
+func (d *Deque) PushFront(idx uint64, item interface{}) {
+	d.Items = append([]Pair{{item, idx}}, d.Items...)
 }
-func (h *maxMinHeap) Swap(i, j int) { h.data[i], h.data[j] = h.data[j], h.data[i] }
 
-func (h *maxMinHeap) Push(x interface{}) {
-	h.data = append(h.data, x)
-}
-func (h *maxMinHeap) Pop() interface{} {
-	old := h.data
-	n := len(old)
-	x := old[n-1]
-	h.data = old[0 : n-1]
-	return x
+// PushBack pushes idx and item(wrapped in Pair) to the end of Deque
+func (d *Deque) PushBack(idx uint64, item interface{}) {
+	d.Items = append(d.Items, Pair{item, idx})
 }
 
-func (h *maxMinHeap) Reset() {
-	h.data = h.data[:0]
-	h.varSet = make(map[interface{}]int64)
-}
-func (h *maxMinHeap) Append(val interface{}) {
-	h.varSet[val]++
-	if h.varSet[val] == 1 {
-		heap.Push(h, val)
+// PopFront pops an item from the front of Deque
+func (d *Deque) PopFront() (interface{}, error) {
+	if len(d.Items) <= 0 {
+		return nil, errors.New("Pop front when deque is empty")
 	}
+	defer func() {
+		d.Items = d.Items[1:]
+	}()
+	return d.Items[0], nil
 }
-func (h *maxMinHeap) Remove(val interface{}) {
-	if h.varSet[val] > 0 {
-		h.varSet[val]--
-	} else {
-		panic("remove a not exist value")
+
+// PopBack pops an item from the end of Deque
+func (d *Deque) PopBack() (interface{}, error) {
+	i := len(d.Items) - 1
+	if i < 0 {
+		return nil, errors.New("Pop back when deque is empty")
 	}
+	defer func() {
```

Comment:
I think using defer here will make the performance worse. 

```
// Benchmark1-8   	227138541	         4.99 ns/op
func Benchmark1(b *testing.B) {
	for i := 0; i < b.N; i++ {
		a := []int{1, 2, 3}
		func() {
			defer func() {
				a = a[1:]
			}()
			_ = a[0]
		}()
	}
}

// Benchmark2-8   	1000000000	         0.310 ns/op
func Benchmark2(b *testing.B) {
	for i := 0; i < b.N; i++ {
		a := []int{1, 2, 3}
		func() {
			_ = a[0]
			a = a[1:]
		}()
	}
}
```

Reply:
Never know that defer will slow this much performance, I'll remove defer.